### PR TITLE
gh-85747: Active voice & suggested edits, 'running/stopping loop' & 'callbacks' subsections of asyncio-eventloop.rst

### PR DESCRIPTION
**Pull request by bskinn**
Another docs revision PR of `asyncio-eventloop.rst` per #85747. 

There are a couple of spots here where I think more explanation than was already present will be useful. I've attempted explanations of my own, but they may be incorrect. I'll mark these with review comments shortly.

I didn't change much of the passive voice related to callback interactions, within the description text of methods that work with callbacks. All the attempts I made to cast those descriptions in active voice seemed to degrade readability.

* Issue: gh-85747

### CODE REVIEW

File: Doc/library/asyncio-eventloop.rst
Comment (bskinn) [inline Doc/library/asyncio-eventloop.rst]: One spot where I think more explanation would be beneficial: explaining *why* "there is no need to call this function when `asyncio.run()` is used." Please correct my guess at the "why", if needed.

File: Doc/library/asyncio-eventloop.rst
Comment (bskinn) [inline Doc/library/asyncio-eventloop.rst]: It wasn't clear to me why it was important to note that 'each callback will be called exactly once' -- if I register a callback once, I would expect it only to be run once. 

My guess here is that a callback could be registered multiple times before execution loops back around to it, and in this case it's still called only once?

Alternatively -- I could see a design decision where a callback, once registered, is called repeatedly as the event loop cycles around, up until the point where it is _de-registered_ by some mechanism... but that would be a more surprising design to me.

Regardless: I suggest more clarity here around this "will be called exactly once".

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [inline Doc/library/asyncio-eventloop.rst]:
```suggestion
   the threads in the :class:`~concurrent.futures.ThreadPoolExecutor`.
   After calling this method,
   :meth:`loop.run_in_executor` will raise a :exc:`RuntimeError` if called
```

Fix broken reference to `ThreadPoolExecutor` (while still displaying it as such)

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [inline Doc/library/asyncio-eventloop.rst]:
```suggestion
   *timeout* specifies the duration in seconds (as a :class:`float`)
   this method gives the executor to finish joining.
   With the default, ``None``, the
```

* Add key clarifying detail that timeout is in seconds as a `float`, which is never stated, and I had to dig into [the code](https://github.com/python/cpython/blob/5d09d11aa0b89aeba187f4f520728ccaf4fc5ac1/Lib/threading.py#L1094) to be sure. This leaves the reader to guess—Float seconds? Integer microseconds? `datetime` objects? `time` structs? Per Diataxis, there should be no doubt or ambiguity in reference.
* Be more concise, direct and to the point, as appropriate for reference docs

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [inline Doc/library/asyncio-eventloop.rst]:
```suggestion
   If the *timeout* is reached, this method emits a :exc:`RuntimeWarning` and
```

Explicitly specify and cross-reference _what_ warning is raised (extracted from the code), if users wanted to `except` it, filter it, check for it, log it, etc).

Also, format the parameter name accordingly

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [inline Doc/library/asyncio-eventloop.rst]:
```suggestion
   .. note:: There is no need to call this method when :func:`asyncio.run` is
      used, because that high-level function handles default executor shutdown
      automatically.
```

* Make this an actual note
* This is a method rather than a function, and calling it thus avoids repetition and potential confusion

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [inline Doc/library/asyncio-eventloop.rst]: It was not totally clear to me whether this means to say that:
* This method must be used If _the callback_ will be run in another thread than _this method is called in_,, which is what I _think_ it is trying to say, or
* This method _must be used from another thread_ to schedule callbacks, as I first read it, or
* Something else?

Here's a suggestion trying to clarify the first option:

```suggestion
   A thread-safe variant of :meth:`call_soon`.
   This method **must** be used instead of :meth:`!call_soon` to
   schedule callbacks if they are to be called from another thread.
```

Also, used bold instead of italics to avoid confusion with the standard formatting of parameter names.

**Review by CAM-Gerlach — COMMENTED** [review]
I've added some clarifications on some further points of confusion (that I had to look up in the code, or guess on), as well as refined the phrasing and fixed a few other issues. Thanks!

**Comment by gvanrossum** [comment]
I'm unsubscribing for now, but do ping me if you need me to arbitrate some choice!

File: Doc/library/asyncio-eventloop.rst
Comment (bskinn) [reply Doc/library/asyncio-eventloop.rst]: @gvanrossum Is this "why" accurate?

File: Doc/library/asyncio-eventloop.rst
Comment (bskinn) [reply Doc/library/asyncio-eventloop.rst]: @gvanrossum Can you provide guidance on why this 'will be called exactly once' language is here?

File: Doc/library/asyncio-eventloop.rst
Comment (bskinn) [reply Doc/library/asyncio-eventloop.rst]: @CAM-Gerlach Is it common reST syntax to have the note text start on the same line as the `.. note::` directive? I'm pretty sure I've usually seen it always as:

```
.. note::

   There is no need ...
   ...
   ...
```

File: Doc/library/asyncio-eventloop.rst
Comment (bskinn) [reply Doc/library/asyncio-eventloop.rst]: @gvanrossum Can you provide clarity on the technical details here?

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [reply Doc/library/asyncio-eventloop.rst]: Stylistically, I typically do

```suggestion
   .. note::
   
      There is no need to call this method when :func:`asyncio.run` is
      used, because that high-level function handles default executor shutdown
      automatically.
```

But either way produces identical rendered output, so long as the directive is configured accordingly (as the default admonitions are).

File: Doc/library/asyncio-eventloop.rst
Comment (gvanrossum) [reply Doc/library/asyncio-eventloop.rst]: I don't recall, it's always been there even in the earliest version of the asyncio docs I can find.

Your addition "even if it is registered multiple times" is incorrect though -- if I register the exact same function twice, it will be called twice.

I recommend leaving this paragraph alone.

Comment (gvanrossum) [reply Doc/library/asyncio-eventloop.rst]: @bskinn
> if I register a callback once, I would expect it only to be run once.

That depends on the type of callback -- it is true here, but out of context that's a faulty assumption. E.g. `add_reader()` adds a callback that may be called multiple times.

File: Doc/library/asyncio-eventloop.rst
Comment (gvanrossum) [reply Doc/library/asyncio-eventloop.rst]: Yeah, it's just because `asyncio.run()` calls various shutdown methods.

File: Doc/library/asyncio-eventloop.rst
Comment (gvanrossum) [reply Doc/library/asyncio-eventloop.rst]: I'd simply reverse it: "When scheduling callbacks from another thread, this function must be called, since `call_soon()` is not thread-safe."

Comment (CAM-Gerlach) [reply Doc/library/asyncio-eventloop.rst]: Taking the above and clarifying further the finer points a bit, specifically what "from another thread" refers to (another thread than the caller? The callback is executed in? Or something else?) and being more explicit that it must be used _instead of_ `call_soon`, both as above, we have


```suggestion
   A thread-safe variant of :meth:`call_soon`.
   When scheduling callbacks to be called from another thread,
   this method **must** be used instead of the non-thread-safe :meth:`!call_soon`.
```

Is this right?

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [inline Doc/library/asyncio-eventloop.rst]:
```suggestion
   .. note::

      There is no need to call this method when :func:`asyncio.run` is
      used, because that high-level function handles default executor shutdown
      automatically.
```

Seems like some trailing white space crept in to my previous example suggestion

File: Doc/library/asyncio-eventloop.rst
Comment (gvanrossum) [reply Doc/library/asyncio-eventloop.rst]: Yeah, though I'm less excited than CAM about the **bold**. (I feel that **bold** for emphasis is pretty extreme, and I don't think there's confusion possible with other uses of italics here.

File: Doc/library/asyncio-eventloop.rst
Comment (gvanrossum) [inline Doc/library/asyncio-eventloop.rst]: Call me old-fashioned, but I don't like starting a sentence with a lowercase letter, even if it's part of an identifier, and I don't like capitalizing the name either, so I liked the original phrasing:
```suggestion
   The *timeout* parameter specifies the duration in seconds (as a :class:`float`)
```

Comment (CAM-Gerlach) [reply Doc/library/asyncio-eventloop.rst]: Yeah, I don't like it either, but on the other hand it was a little more concise and direct, as appropriate for an API reference per Diataxis. It just comes down which you want to prioritize...

Comment (gvanrossum) [reply Doc/library/asyncio-eventloop.rst]: @bskinn can choose then.

File: Doc/library/asyncio-eventloop.rst
Comment (bskinn) [reply Doc/library/asyncio-eventloop.rst]: Another option might be [RFC2199 "MUST"](https://www.rfc-editor.org/rfc/rfc2119)?

A quick search of the codebase shows what appear to be numerous uses of RFC2119, so this seems stylistically consistent to me.

File: Doc/library/asyncio-eventloop.rst
Comment (bskinn) [reply Doc/library/asyncio-eventloop.rst]: This change reverted entire by 7d18a15

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [reply Doc/library/asyncio-eventloop.rst]: To me that feels even stronger than the bold. If bold is too strong, then we can just use italics anyway; certainly not worth holding up this PR over:

```suggestion
   A thread-safe variant of :meth:`call_soon`.
   When scheduling callbacks to be called from another thread,
   this method *must* be used instead of the non-thread-safe :meth:`!call_soon`.
```

Comment (bskinn) [reply Doc/library/asyncio-eventloop.rst]: Not to strain overly hard at a gnat, but --- to me, the emphasis of **bold** has a different meaning from the emphasis of MUST: Bold is a semantic emphasis, as part of an attempt to persuade the reader by reasoning; whereas MUST is a technical emphasis, indicating how one is to use the method correctly.

Given that it is critical in a technical sense for a caller to use this method and not `call_soon()` from another thread, I think the emphasis of MUST is specifically warranted here.

Comment (CAM-Gerlach) [reply Doc/library/asyncio-eventloop.rst]: That's true, but I at least strongly (heh) associate it with mandatory clauses in normative specifications, as opposed to user-facing instructions. But...meh, it's not _that_ important to me what color we paint the bikeshed... :)

Comment (bskinn) [reply Doc/library/asyncio-eventloop.rst]: I appreciate your perspective, it's what I was going for more than anything. (Apologies to anyone watching the PR for the noise as we discussed.)

On looking back at Guido's original recommendation, I think his wording is semantically different from your recommendation, so I'm actually going to edit to match his, since he is the subject matter expert here. Commit coming shortly.

**Comment by bskinn** [comment]
@gvanrossum -- I believe this PR is ready for consideration for merge. Thank you for the input & guidance!

**Review by gvanrossum — CHANGES_REQUESTED** [review]
I'm sorry, but there are still too many grammatical oddities to approve the PR. Please ask one of the docs folks (e.g. @CAM-Gerlach) to go over the grammar carefully.

**Comment by arhadthedev** [comment]
> Please ask one of the docs folks [...] to go over the grammar carefully.

@python/proofreaders

**Comment by bskinn** [comment]
@gvanrossum I'm on proofreaders with @CAM-Gerlach. 

The goal of the PR was the passive->active voice conversion, per the linked issue. I didn't realize you wanted a general grammar review also for these.

Happy to make another pass, with @CAM-Gerlach if he desires.

**Comment by gvanrossum** [comment]
Well, I don’t like the result to read worse than the original. :-)

**Comment by bskinn** [comment]
Oh, there's work needed on the changes themselves. Got it - sorry!

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [inline Doc/library/asyncio-eventloop.rst]:
```suggestion
   Once this method has been called,
   using the default executor with :meth:`loop.run_in_executor`
   will raise a :exc:`RuntimeError`.
```

This reads a fair bit more clearly to me, as the grammar and flow of ideas is more straightforward for me to parse.

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [inline Doc/library/asyncio-eventloop.rst]:
```suggestion
   The *timeout* parameter specifies the amount of time
   (in :class:`float` seconds) the executor will be given to finish joining.
```

I had trouble getting active voice to not sound awkward here, and I don't think it really adds much to reader understanding here—as a parameter description in an the API reference, the method being described is the clear actor in this case. Therefore, I reverted to the original wording, with the key clarification that the time given is in (floating point) seconds.

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [inline Doc/library/asyncio-eventloop.rst]:
```suggestion
   With the default, ``None``,
   the executor is allowed an unlimited amount of time.
```

Likewise, here, I feel inserting "the method" here makes the sentence structure more complex for the reader to parse without really aiding meaning. Here, the exact implementation detail doing the allowing (the subject) is a lot less important than the object being allowed (the executor), which is exactly what passive voice is useful for. And the method is explicitly what we are describing here, which the basic structure of API reference documentation already makes clear, and repeatedly restating is not really in line with the concision and directness that serves reference-type docs well.

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [inline Doc/library/asyncio-eventloop.rst]:
```suggestion
   If the *timeout* is reached, a :exc:`RuntimeWarning` is emitted
   and the default executor is terminated
   without waiting for its threads to finish joining.
```

Like before, reading and thinking over this again, passive voice is generally idiomatic and keeping with the Diataxis principles in API reference documentation in the case of referring to the construct that is explicitly stated as being described, and avoids implying implementation details that may or may not be entirely correct (e.g. it implies the method directly emits the `RuntimeWarning`, when it may in fact be emitted deeper in the execution stack).

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [inline Doc/library/asyncio-eventloop.rst]:
```suggestion
      There is no need to call this method when :func:`asyncio.run` is used,
      because that function handles default executor shutdown automatically.
```

You might want to consider eliding "high-level" to improve readability here and avoid sounding a bit awkward.

**Review by CAM-Gerlach — COMMENTED** [review]
After some more consideration and thought, various suggestions to improve the readability and clarity of the grammar and phrasing and avoid regressions on that front.

**Comment by terryjreedy** [comment]
I am taking a look,

File: Doc/library/asyncio-eventloop.rst
Comment (terryjreedy) [reply Doc/library/asyncio-eventloop.rst]: I prefer CAM version.

File: Doc/library/asyncio-eventloop.rst
Comment (terryjreedy) [reply Doc/library/asyncio-eventloop.rst]: Float seconds is critical.  I might replace 'will be given' to 'will have', which removes the passive.

File: Doc/library/asyncio-eventloop.rst
Comment (terryjreedy) [reply Doc/library/asyncio-eventloop.rst]: or 'the executor has ....

File: Doc/library/asyncio-eventloop.rst
Comment (terryjreedy) [reply Doc/library/asyncio-eventloop.rst]: I was wondering about the implied implementation detail.  agree with leave it out.

File: Doc/library/asyncio-eventloop.rst
Comment (terryjreedy) [inline Doc/library/asyncio-eventloop.rst]:
```suggestion
   The optional keyword-only *context* argument specifies a
   custom :class:`contextvars.Context` for the *callback* to run in.
```

File: Doc/library/asyncio-eventloop.rst
Comment (terryjreedy) [inline Doc/library/asyncio-eventloop.rst]:
```suggestion
   Return an instance of :class:`asyncio.Handle`, which can be
```

File: Doc/library/asyncio-eventloop.rst
Comment (terryjreedy) [inline Doc/library/asyncio-eventloop.rst]:
```suggestion
      Do not call this method when using :func:`asyncio.run`.  The latter handles default executor shutdown
      automatically.
```

**Review by terryjreedy — CHANGES_REQUESTED** [review]

File: Doc/library/asyncio-eventloop.rst
Comment (terryjreedy) [reply Doc/library/asyncio-eventloop.rst]: See my alternate suggestion in my review.

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [reply Doc/library/asyncio-eventloop.rst]: We might, but the current framing is consistent with this being an external, discretionary timeout acting on the executor, as opposed to a requirement intrinsic to the executor or its context.

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [reply Doc/library/asyncio-eventloop.rst]: Like the above, that framing seems more appropriate for an intrinsic requirement than an externally-set  discretionary deadline. But if we do change the above, we should presumably change this to match

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [reply Doc/library/asyncio-eventloop.rst]: Yeah, I think I like a slight variant of yours better.

Just FYI, you can also reply with a suggestion of your own in the same thread, to keep things in one place.

Comment (CAM-Gerlach) [reply Doc/library/asyncio-eventloop.rst]: I was trying to avoid "the latter" to avoid any confusion, but I think it works okay and flows better overall merged into a single sentence:

```suggestion
      Do not call this method when using :func:`asyncio.run`,
      as the latter handles default executor shutdown automatically.
```

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [reply Doc/library/asyncio-eventloop.rst]: Particularly if we frame it this way, this sentence/para would seem to fit best immediately after the first/summary line. This also follows the loose convention used other places in this module (e.g. in the following `call_later` method) and in other modules.

**Comment by bskinn** [comment]
@CAM-Gerlach, it seems like you and @terryjreedy are in a good flow for further polishing here.

Do you have write access on my fork? If not, do you want it? I don't want my pace of review to hold up applying changes to the PR branch.

File: Doc/library/asyncio-eventloop.rst
Comment (CAM-Gerlach) [reply Doc/library/asyncio-eventloop.rst]: Applied the minor refinement of your suggestion instead

**Comment by bskinn** [comment]
> Just FYI, you can just batch-apply all desired review comments in one go by clicking on the [`Files changed`](https://github.com/python/REPO/pull/NUMBER/files), clicking Add to batch on each suggestion, and finally pressing Commit.

Yup, I know -- super handy feature!  I wasn't sure which suggestions were 'final', or if there was more discussion coming...so, I didn't feel comfortable committing anything without a thorough review that I couldn't be sure when I could carry out.

**Review by CAM-Gerlach — APPROVED** [review]
LGTM from my end, now

**Pull request merged**
